blogs/qserial.md

```diff
@@ -0,0 +1,243 @@
+          nextPromise = success();
+          if (!isPromiseLike(nextPromise)) {
+            error.message = "Task " + key + " did not return a promise.";
+            throw error;
+          }
+        } else {
+          //Wait until the previous promise has resolved or rejected to execute the next task
+          nextPromise = prevPromise.then(
+            /*success*/function (data) {
+              if (!success) { return data; }
+              var ret = success(data);
+              if (!isPromiseLike(ret)) {
+                error.message = "Task " + key + " did not return a promise.";
+                throw error;
+              }
+              return ret;
+            },
+            /*failure*/function (reason) {
+              if (!fail) { return $delegate.reject(reason); }
+              var ret = fail(reason);
+              if (!isPromiseLike(ret)) {
+                error.message = "Fail for task " + key + " did not return a promise.";
+                throw error;
+              }
+              return ret;
+            },
+            notify);
+        }
+        prevPromise = nextPromise;
+      });
+
+      return prevPromise;
+    }
+
+    $delegate.serial = serial;
+    return $delegate;
+  });
+});
+</pre>
+<br/>
+Now that we have the full $q.serial API, let's investigate its power.  $q.serial is most useful when you have a tasks that are expensive, tie up resources, and are not time sensitive.
+
+At <a href="http://hurdlr.com/">Hurdlr</a> we encountered this exact criteria the other day.  We build a mobile app.  Internet tends to be less reliable on phones than on computers.  So we try to minimize server calls.  
+
+One way we minimize server calls is by eagerly loading bulk data.  So depending on module, user's config, etc. we execute a set of tasks that load bulk data.  We don't want to execute all the tasks at once because it can tie up more database resources and phone resources when caching than we'd like.
+ 
+<pre>
+var tasks = [];
+if (user.config.foo) {
+    tasks.push(retrieveFooData);
+}
+if (module.bar) {
+    tasks.push(retrieveBarData);
+}
+$q.serial(tasks);
+</pre>
+<br/>
+Ok, so we've got 99% of the functionality we need.  But I should mention that $q.serial supports reject and notify also.  Oh, and $q.serial can accept a hash instead of an array.
+
+<pre>
+var task1 = function () {
+  var deferred = $q.defer();
+  console.log("one");
+  $timeout(function () { deferred.notify(); });
+  $timeout(function () { deferred.reject(); }, 5000);
+  return deferred.promise;
+};
+var task2 = {
+  success: function () {
+    console.log("two");
+    return $q.when();
+  }, fail: function () {
+    console.log("fail two");
+    return $q.reject();
+  }, notify: function () {
+    console.log("notified two");
+  }
+};
+
+$q.serial({ one: task1, two: task2 });
+</pre>
+<pre>
+one
+notified two
+//5 seconds
+fail two
+</pre>
+<br/>
+
+So that's $q.serial, a generalized way to execute async tasks serially in AngularJS.
```

Comment:
$q is a game changer.  It's based off of https://github.com/kriskowal/q.  I suspect that library is very light weight.  It's a superset of $q and seems really amazing.

I think promises would be huge for APT code cleanliness.  Also, using underscore and Angular/Ember would be a huge productivity boost.  My javascript development is so much faster and more elegant because of $q and AngularJS.